### PR TITLE
QUIC: Harden ring buffer against internal misuse

### PR DESCRIPTION
**Pull request by hlandau**
Fixes https://github.com/openssl/openssl/issues/21892

### CODE REVIEW

**Review by paulidale — APPROVED** [review]

**Comment by paulidale** [comment]
Honestly not sure we need this, but it shouldn't hurt the extra arithmetic will amount to basically nothing on most machines.

File: include/internal/ring_buf.h [hunk 1]
```diff
@@ -113,6 +122,9 @@ static ossl_inline size_t ring_buf_push(struct ring_buf *r,
         if (buf_len > avail)
             buf_len = avail;
 
+        if (buf_len > MAX_OFFSET - r->head_offset)
+            buf_len = (size_t)(MAX_OFFSET - r->head_offset);
```
[the remark this answers is not in the document]
Comment (hlandau) [reply include/internal/ring_buf.h:126]: We don't allocate anywhere near that much &mdash; obviously we never allocate `1<<62` bytes of memory. But the `head_offset` field tracks the logical offset in a QUIC stream &mdash; so we might have a 4k ring buffer but where the offsets have reached e.g. `1<<60` etc. A QUIC stream can't be used further once it reaches the limit, so this limitation aligns perfectly with QUIC.

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
@paulidale please reconfirm

**Review by mattcaswell — APPROVED** [review]

**Review by paulidale — APPROVED** [review]

**Comment by hlandau** [comment]
Merged to master. Thanks for the reviews.

**Pull request closed without merging**